src/impl/rocshmemspace/Kokkos_ROCSHMEMSpace.hpp

```diff
@@ -0,0 +1,154 @@
+//
+//@HEADER
+
+#ifndef KOKKOS_ROCSHMEMSPACE_HPP
+#define KOKKOS_ROCSHMEMSPACE_HPP
+
+#include <cstring>
+#include <iosfwd>
+#include <string>
+#include <typeinfo>
+
+#include <Kokkos_Core.hpp>
+
+#include <Kokkos_RemoteSpaces.hpp>
+#include <mpi.h>
+/*--------------------------------------------------------------------------*/
+
+namespace Kokkos {
+namespace Experimental {
+
+class RemoteSpaceSpecializeTag {};
+
+class ROCSHMEMSpace {
+ public:
+#if defined(KOKKOS_ENABLE_HIP)
+  using execution_space = Kokkos::HIP;
+#else
+#error \
+    "At least the following device execution space must be defined: Kokkos::HIP."
+#endif
+  using memory_space = ROCSHMEMSpace;
+  using device_type  = Kokkos::Device<execution_space, memory_space>;
+  using size_type    = size_t;
+
+  ROCSHMEMSpace();
+  ROCSHMEMSpace(ROCSHMEMSpace &&rhs)      = default;
+  ROCSHMEMSpace(const ROCSHMEMSpace &rhs) = default;
+  ROCSHMEMSpace &operator=(ROCSHMEMSpace &&) = default;
+  ROCSHMEMSpace &operator=(const ROCSHMEMSpace &) = default;
+  ~ROCSHMEMSpace()                                = default;
+
+  explicit ROCSHMEMSpace(const MPI_Comm &);
+
+  void *allocate(const size_t arg_alloc_size) const;
+
+  void deallocate(void *const arg_alloc_ptr, const size_t arg_alloc_size) const;
+
+  void *allocate(const int *gids, const int &arg_local_alloc_size) const;
+
+  void deallocate(const int *gids, void *const arg_alloc_ptr,
+                  const size_t arg_alloc_size) const;
+
+  /**\brief Return Name of the MemorySpace */
+  static constexpr const char *name() { return m_name; }
+
+  void fence();
+
+  int allocation_mode;
+  int64_t extent;
+
+  void impl_set_allocation_mode(const int);
+  void impl_set_extent(int64_t N);
+
+ private:
+  static constexpr const char *m_name = "ROCSHMEM";
+  friend class Kokkos::Impl::SharedAllocationRecord<
+      Kokkos::Experimental::ROCSHMEMSpace, void>;
+};
+
+KOKKOS_FUNCTION
+size_t get_num_pes();
+KOKKOS_FUNCTION
+size_t get_my_pe();
+KOKKOS_FUNCTION
+size_t get_indexing_block_size(size_t size);
+std::pair<size_t, size_t> getRange(size_t size, size_t pe);
+
+}  // namespace Experimental
+}  // namespace Kokkos
+
+namespace Kokkos {
+namespace Impl {
+
+template <>
+struct DeepCopy<HostSpace, Kokkos::Experimental::ROCSHMEMSpace> {
+  DeepCopy(void *dst, const void *src, size_t);
+};
+
+template <>
+struct DeepCopy<Kokkos::Experimental::ROCSHMEMSpace, HostSpace> {
+  DeepCopy(void *dst, const void *src, size_t);
+};
+
+template <class ExecutionSpace>
+struct DeepCopy<Kokkos::Experimental::ROCSHMEMSpace,
+                Kokkos::Experimental::ROCSHMEMSpace, ExecutionSpace> {
+  DeepCopy(void *dst, const void *src, size_t n);
+  DeepCopy(const ExecutionSpace &exec, void *dst, const void *src, size_t n);
+};
+
+template <>
+struct MemorySpaceAccess<Kokkos::Experimental::ROCSHMEMSpace,
+                         Kokkos::Experimental::ROCSHMEMSpace> {
+  enum { assignable = true };
+  enum { accessible = true };
+  enum { deepcopy = false };
```

Comment:
We can. This is an artifact from other spaces and we'd need to update there as well. In case the spaces are accessible the behavior would be as local_deep_copy.

Reply:
For this PR I think we can leave it as it and fix all spaces at once.